From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Lukas Voegl <lvoegl@tdt.de>
Date: Thu, 17 Sep 2026 11:26:51 +0200
Subject: [PATCH 3/6] meig: implement MeiG ASR plugin

Implements a plugin for ASR based MeiG modems using shared MeiG utilities.

Signed-off-by: Lukas Voegl <lvoegl@tdt.de>
---
 meson.build                                  |  1 +
 meson_options.txt                            |  1 +
 src/plugins/meig/77-mm-meig-port-types.rules | 20 +++++
 src/plugins/meig/mm-plugin-meig-asr.c        | 80 ++++++++++++++++++++
 src/plugins/meson.build                      | 15 ++++
 src/plugins/mm-builtin-plugins.c             |  6 ++
 src/plugins/tests/test-udev-rules.c          | 11 +++
 7 files changed, 134 insertions(+)
 create mode 100644 src/plugins/meig/77-mm-meig-port-types.rules
 create mode 100644 src/plugins/meig/mm-plugin-meig-asr.c

diff --git a/meson.build b/meson.build
index dbcde306..59dd9bf5 100644
--- a/meson.build
+++ b/meson.build
@@ -345,6 +345,7 @@ plugins_options_reqs = {
   'linktop': {'available': true, 'shared': []},
   'longcheer': {'available': true, 'shared': []},
   'mbm': {'available': true, 'shared': []},
+  'meig-asr': {'available': true, 'shared': ['meig', 'asr']},
   'motorola': {'available': true, 'shared': []},
   'mtk-legacy': {'available': true, 'shared': mtk_shared_reqs},
   'mtk': {'available': true, 'shared': mtk_shared_reqs},
diff --git a/meson_options.txt b/meson_options.txt
index c0609fe8..9bf7f344 100644
--- a/meson_options.txt
+++ b/meson_options.txt
@@ -44,6 +44,7 @@ option('plugin_iridium', type: 'feature', value: 'auto', description: 'enable ir
 option('plugin_linktop', type: 'feature', value: 'auto', description: 'enable linktop plugin support')
 option('plugin_longcheer', type: 'feature', value: 'auto', description: 'enable longcheer plugin support')
 option('plugin_mbm', type: 'feature', value: 'auto', description: 'enable mbm plugin support')
+option('plugin_meig_asr', type: 'feature', value: 'auto', description: 'enable meig asr plugin support')
 option('plugin_motorola', type: 'feature', value: 'auto', description: 'enable motorola plugin support')
 option('plugin_mtk_legacy', type: 'feature', value: 'auto', description: 'enable mtk legacy plugin support')
 option('plugin_mtk', type: 'feature', value: 'auto', description: 'enable mtk plugin support')
diff --git a/src/plugins/meig/77-mm-meig-port-types.rules b/src/plugins/meig/77-mm-meig-port-types.rules
new file mode 100644
index 00000000..f32dfa41
--- /dev/null
+++ b/src/plugins/meig/77-mm-meig-port-types.rules
@@ -0,0 +1,20 @@
+# do not edit this file, it will be overwritten on update
+
+ACTION!="add|change|move|bind", GOTO="mm_meig_end"
+SUBSYSTEMS=="usb", ATTRS{idVendor}=="2dee", GOTO="mm_meig_start"
+GOTO="mm_meig_end"
+
+LABEL="mm_meig_start"
+SUBSYSTEMS=="usb", ATTRS{bInterfaceNumber}=="?*", ENV{.MM_USBIFNUM}="$attr{bInterfaceNumber}"
+
+# MeiG SLM770A (ECM)
+#  ttyUSB0 (if #2): ASR DIAG port (ignore for now)
+#  ttyUSB1 (if #3): AT port
+#  ttyUSB2 (if #4): AT port
+#  ttyUSB3 (if #5): GPS data port
+ATTRS{idVendor}=="2dee", ATTRS{idProduct}=="4d58", ENV{.MM_USBIFNUM}=="02", SUBSYSTEM=="tty", ENV{ID_MM_PORT_IGNORE}="1"
+ATTRS{idVendor}=="2dee", ATTRS{idProduct}=="4d58", ENV{.MM_USBIFNUM}=="03", SUBSYSTEM=="tty", ENV{ID_MM_PORT_TYPE_AT_PRIMARY}="1"
+ATTRS{idVendor}=="2dee", ATTRS{idProduct}=="4d58", ENV{.MM_USBIFNUM}=="04", SUBSYSTEM=="tty", ENV{ID_MM_PORT_TYPE_AT_SECONDARY}="1"
+ATTRS{idVendor}=="2dee", ATTRS{idProduct}=="4d58", ENV{.MM_USBIFNUM}=="05", SUBSYSTEM=="tty", ENV{ID_MM_PORT_GPS}="1"
+
+LABEL="mm_meig_end"
diff --git a/src/plugins/meig/mm-plugin-meig-asr.c b/src/plugins/meig/mm-plugin-meig-asr.c
new file mode 100644
index 00000000..70c3e45d
--- /dev/null
+++ b/src/plugins/meig/mm-plugin-meig-asr.c
@@ -0,0 +1,80 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details:
+ *
+ * Copyright (C) 2026 TDT AG
+ */
+
+#include "mm-plugin.h"
+#include <gmodule.h>
+
+#define _LIBMM_INSIDE_MM
+#include <libmm-glib.h>
+
+#include "mm-log-object.h"
+#include "mm-broadband-modem-meig-asr.h"
+#include "mm-modem-helpers-meig.h"
+#include "mm-plugin-common.h"
+
+#define MM_TYPE_PLUGIN_MEIG_ASR mm_plugin_meig_asr_get_type ()
+MM_DEFINE_PLUGIN (MEIG_ASR, meig_asr, MeigAsr)
+
+static MMBaseModem *
+create_modem (MMPlugin     *self,
+              const gchar  *uid,
+              const gchar  *physdev,
+              const gchar **drivers,
+              guint16       vendor,
+              guint16       product,
+              guint16       subsystem_vendor,
+              guint16       subsystem_device,
+              GList        *probes,
+              GError      **error)
+{
+    mm_obj_dbg (self, "ASR based MeiG modem found");
+    return MM_BASE_MODEM (mm_broadband_modem_meig_asr_new (uid,
+                                                           physdev,
+                                                           drivers,
+                                                           mm_plugin_get_name (self),
+                                                           vendor,
+                                                           product));
+}
+
+MM_PLUGIN_NAMED_CREATOR_SCOPE MMPlugin *
+mm_plugin_create_meig_asr (void)
+{
+    static const gchar          *subsystems[] = { "tty", "net", NULL };
+    static const mm_uint16_pair  products[]   = {
+        { 0x2dee, 0x4d58 }, /* MeiG SLM770A */
+        { 0, 0 }
+    };
+
+    return MM_PLUGIN (
+        g_object_new (MM_TYPE_PLUGIN_MEIG_ASR,
+                      MM_PLUGIN_NAME,                MM_MODULE_NAME,
+                      MM_PLUGIN_ALLOWED_SUBSYSTEMS,  subsystems,
+                      MM_PLUGIN_ALLOWED_PRODUCT_IDS, products,
+                      MM_PLUGIN_ALLOWED_AT,          TRUE,
+                      NULL));
+}
+
+static void
+mm_plugin_meig_asr_init (MMPluginMeigAsr *self)
+{
+}
+
+static void
+mm_plugin_meig_asr_class_init (MMPluginMeigAsrClass *klass)
+{
+    MMPluginClass *plugin_class = MM_PLUGIN_CLASS (klass);
+
+    plugin_class->create_modem = create_modem;
+}
diff --git a/src/plugins/meson.build b/src/plugins/meson.build
index 789a25bf..8248eea0 100644
--- a/src/plugins/meson.build
+++ b/src/plugins/meson.build
@@ -712,6 +712,21 @@ if plugins_options['mbm']
   plugins_udev_rules += files('mbm/77-mm-ericsson-mbm.rules')
 endif
 
+# plugin: meig (asr)
+if plugins_options['meig-asr']
+  test_udev_rules_dir_c_args = ['-DTESTUDEVRULESDIR_MEIG_ASR="@0@"'.format(plugins_dir / 'meig')]
+  plugins_test_udev_rules_dir_c_args += test_udev_rules_dir_c_args
+
+  incs = plugins_incs + [meig_inc] + [asr_inc]
+
+  plugins += {'plugin-meig-asr': {
+    'plugin': true,
+    'module': {'sources': files('meig/mm-plugin-meig-asr.c'), 'include_directories': incs, 'c_args': test_udev_rules_dir_c_args + ['-DMM_MODULE_NAME="meig-asr"']},
+  }}
+
+  plugins_udev_rules += files('meig/77-mm-meig-port-types.rules')
+endif
+
 # plugin: motorola
 if plugins_options['motorola']
   sources = files(
diff --git a/src/plugins/mm-builtin-plugins.c b/src/plugins/mm-builtin-plugins.c
index c7f91633..ee555c31 100644
--- a/src/plugins/mm-builtin-plugins.c
+++ b/src/plugins/mm-builtin-plugins.c
@@ -73,6 +73,9 @@ MMPlugin *mm_plugin_create_longcheer (void);
 #if defined ENABLE_PLUGIN_MBM
 MMPlugin *mm_plugin_create_mbm (void);
 #endif
+#if defined ENABLE_PLUGIN_MEIG_ASR
+MMPlugin *mm_plugin_create_meig_asr (void);
+#endif
 #if defined ENABLE_PLUGIN_MOTOROLA
 MMPlugin *mm_plugin_create_motorola (void);
 #endif
@@ -211,6 +214,9 @@ mm_builtin_plugins_load (void)
 #if defined ENABLE_PLUGIN_MBM
     PREPEND_PLUGIN (mbm);
 #endif
+#if defined ENABLE_PLUGIN_MEIG_ASR
+    PREPEND_PLUGIN (meig_asr);
+#endif
 #if defined ENABLE_PLUGIN_MOTOROLA
     PREPEND_PLUGIN (motorola);
 #endif
diff --git a/src/plugins/tests/test-udev-rules.c b/src/plugins/tests/test-udev-rules.c
index 8f5ffd64..fe205a44 100644
--- a/src/plugins/tests/test-udev-rules.c
+++ b/src/plugins/tests/test-udev-rules.c
@@ -72,6 +72,14 @@ test_mbm (void)
 }
 #endif
 
+#if defined ENABLE_PLUGIN_MEIG_ASR
+static void
+test_meig_asr (void)
+{
+    common_test (TESTUDEVRULESDIR_MEIG_ASR);
+}
+#endif
+
 #if defined ENABLE_PLUGIN_NOKIA_ICERA
 static void
 test_nokia_icera (void)
@@ -231,6 +239,9 @@ int main (int argc, char **argv)
 #if defined ENABLE_PLUGIN_MBM
     g_test_add_func ("/MM/test-udev-rules/mbm", test_mbm);
 #endif
+#if defined ENABLE_PLUGIN_MEIG_ASR
+    g_test_add_func ("/MM/test-udev-rules/meig-asr", test_meig_asr);
+#endif
 #if defined ENABLE_PLUGIN_NOKIA_ICERA
     g_test_add_func ("/MM/test-udev-rules/nokia-icera", test_nokia_icera);
 #endif
